lib/Requalify.thy

```diff
@@ -1,117 +1,474 @@
+   Note: Locale_Name.fact_name outside the locale resolves to a global name.
+
+   When we are inside a locale, the lookup is more interesting. Supplying "short_name" will result
+   in "local.short_name", which we then need to match to some name in the global context. We do
+   this by going through *all* the fact names in the current context, searching for matches
+   of the form "X.Y.short_name", where we hope X is some theory, and Y is some locale.
 
+   Since "X.Y.short_name" is not sufficiently unique, we must also check that the theorems under
+   the discovered name match the ones under "local.short_name". *)
 fun global_fact ctxt nm =
 let
    val facts = Proof_Context.facts_of ctxt;
    val {name, thms, ...} = (Facts.retrieve (Context.Proof ctxt) facts (nm, Position.none));
 
-   fun tl' (_ :: xs) = xs
-     | tl' _ = []
-
-   fun matches suf (gnm, gthms)  =
-   let
-     val gsuf = Long_Name.explode gnm |> tl' |> tl' |> Long_Name.implode;
-
-   in suf = gsuf andalso eq_list (Thm.equiv_thm (Proof_Context.theory_of ctxt)) (thms, gthms)
-   end
+   fun matches suffix (global_name, global_thms) =
+     suffix = (Long_Name.explode global_name |> tl' |> tl' |> Long_Name.implode)
+     andalso eq_list (Thm.equiv_thm (Proof_Context.theory_of ctxt)) (thms, global_thms)
 in
-  case Long_Name.dest_local name of NONE => (name, thms) | SOME suf =>
-    (case (find_first (matches suf) (all_facts_of ctxt)) of
-       SOME x => x
-     | NONE => raise Fail ("Couldn't find global equivalent of local fact: " ^ nm))
+  case Long_Name.dest_local name of
+    NONE => (name, thms)
+  | SOME suffix =>
+     (case (find_first (matches suffix) (all_facts_of ctxt)) of
+        SOME x => x
+      | NONE => raise Fail ("Couldn't find global equivalent of local fact: " ^ nm))
 end
 
-fun syntax_alias global_alias local_alias b (name : string) =
-  Local_Theory.declaration {syntax = false, pos = Position.none, pervasive = true} (fn phi =>
-    let val b' = Morphism.binding phi b
-    in Context.mapping (global_alias b' name) (local_alias b' name) end);
+val alias = Parse.reserved "aliasing" >> K true
+val alias_default = false
 
-val alias_fact = syntax_alias Global_Theory.alias_fact Proof_Context.alias_fact;
-val const_alias = syntax_alias Sign.const_alias Proof_Context.const_alias;
-val type_alias = syntax_alias Sign.type_alias Proof_Context.type_alias;
+(* (aliasing) only *)
+val generic_options = Scan.optional (Args.parens alias >> (fn x => (x, ""))) (alias_default, "")
 
-in
+(* e.g. ARM, ARM_A, ARM_H *)
+val arch_suffix = ((Parse.reserved "A" || Parse.reserved "H") >> (fn s =>  "_" ^ s))
+fun arch_prefix suffix = getenv_strict "L4V_ARCH" ^ suffix
 
-fun gen_requalify get_proper_nm parse_nm check_nm alias =
-  (Parse.opt_target  --  Scan.repeat1 (Parse.position (Scan.ahead parse_nm -- Parse.name))
-    >> (fn (target,bs) =>
-      Toplevel.local_theory NONE target (fn lthy =>
-      let
+(* ([aliasing][,] [A|H]) in that order *)
```

Comment:
This is because of n+1-flavour soup. If I allow both, we will see someone in the future discovering that they can write `(H, aliasing)` while everyone else writes it the other way around. I'm fine to change that order, but for some reason I got it stuck into my head that I only want one. What do you think?

Reply:
I'm easy. Fixing it to one order does reduce the number of styles, but it also can be frustrating to use if you don't remember what order it should be and are only getting the typical Isabelle "bad syntax" error message.

Would it make sense to try to give an informative error message when the order is wrong?

Reply:
I'll be honest, I do not know how one does that.

Reply:
I have never tried this in Isabelle, but usually, you'd explicitly parse the wrong order and then produce an error message (I guess throw an exception with a useful message?)

Reply:
That seems like a very specific workaround to this situation, at which point I'm wondering if you want me to make the parser deal with arbitrary option order instead. In any case, if you don't remember the exact "aliasing/A/H" phrases, you'll still get
```
Outer syntax error⌂: name expected,
but keyword (⌂ was found
```
so I'm not sure I see the benefit.

Reply:
Can anyone think of other commands or methods with multiple options, so that we can see if there is a standard approach to this? The only one I can immediately think of is `wp`, which I made deal with arbitrary order. One of the options there is a debug one though so it's not quite the same, since there won't be different orders left lying around in the proofs.

Reply:
Can't think of any. Looked at find_theorems, but that one has required order as well.

Reply:
I don't think there is a consistent policy. The datatype command takes in a list of options and plugin filters without order. The sledgehammer_param commands and similar do the same. But I'm sure there are also commands that do require a specific order, at least I remember the frustrations of not remembering order :-).

It's going to be easy to remember for this one, so I don't really mind.

For the parse error: good error messages are always specific to the situation they are for and usual based on very specific code -- that's what makes them informative. In this case we could also make it a warning, actually, i.e. parse a list of options and warn if the list is not in the order we like. We don't really have to do that here, just throwing it out some ideas. (All of these are not very Isabelle things to do anyway, of course).

Reply:
I can give this a go later I guess.

Reply:
The "list of options" thing is also weird to me. I mean, someone could do `(A, aliasing, H)` if it's a list, so the error checking becomes more complex.

Reply:
Or even worse: `(A, H, aliasing, A, A, H, A)` ...

Reply:
If we don't care about order, then it's a case of `sort|uniq` and then error if it's both A and H, but if we care about order I'd need to check for duplicates... I am not getting a clear image of good interface here.

Reply:
I think we'll leave this for another time. Have to redirect resources to arch-splitting.